polyply/src/nonbond_engine.py

```diff
@@ -154,12 +165,24 @@ def remove_positions(self, mol_idx, node_key):
         """
         gndx = self.nodes_to_gndx[(mol_idx, node_key)]
         self.positions[gndx] = np.array([np.inf, np.inf, np.inf])
-        self.defined_idxs[-1].remove(gndx)
-        new_tree = scipy.spatial.ckdtree.cKDTree(self.positions[self.defined_idxs[-1]],
-                                                 boxsize=self.boxsize,
-                                                 balanced_tree=False,
-                                                 compact_nodes=False)
-        self.position_trees[-1] = new_tree
+
+        if gndx in self.gndx_to_tree:
+            tree_idx = self.gndx_to_tree[gndx]
+            self.defined_idxs[tree_idx].remove(gndx)
+            new_tree = scipy.spatial.ckdtree.cKDTree(self.positions[self.defined_idxs[tree_idx]],
+                                                     boxsize=self.boxsize,
+                                                     balanced_tree=False,
+                                                     compact_nodes=False)
+            self.position_trees[tree_idx] = new_tree
+            del self.gndx_to_tree[gndx]
+
+    def remove_molecule_positions(self, mol_idx, molecule):
+        """
+        Set all positions of the nodes in molecule
+        to infinity.
+        """
+        for node in molecule.nodes:
+            self.remove_positions(mol_idx, node)
```

Comment:
Wildly inefficient since you'll be rebuilding trees N times.
I would suggest adapting remove_positions to accept multiple node indices, so you can first wipe the positions themselves (and their indices), and then rebuild the tree(s) once.

Reply:
It is isn't it. I must admit I was quite tired when I played around with this section of the code.